File: interfaces/julia/examples/laplacian.jl
#!/usr/bin/env julia
# Copyright (c) 2024, Lawrence Livermore National Security, LLC.
# See the top-level LICENSE and NOTICE files for details.

using HypreDrive
using LinearAlgebra

mutable struct Params
    input::Union{Nothing,String}
    n::NTuple{3,Int}
    c::NTuple{3,Float64}
    p::NTuple{3,Int}
    stencil::Int
    nsolve::Int
    verbose::Int
    hypredrv_args::Vector{String}
end

function usage()
    println("""
Usage: mpiexec -n <np> julia --project=interfaces/julia interfaces/julia/examples/laplacian.jl [options]

Options:
  -i, --input <file>       YAML configuration file for solver settings
  -a, --args <key value>...  Hypredrive YAML overrides, e.g.
                             -a --solver:pcg:max_iter 100 (must come last)
  -n <nx> <ny> <nz>        Global grid dimensions (default: 10 10 10)
  -c <cx> <cy> <cz>        Diffusion coefficients (default: 1.0 1.0 1.0)
  -P <px> <py> <pz>        Processor grid dimensions (default: 1 1 1)
  -s <val>                 Stencil type: 7, 19, or 27 (default: 7)
  -ns, --nsolve <n>        Number of solves using one assembled matrix (default: 5)
  -v, --verbose <n>        Verbosity bitset (default: 1)
                             0x1: print setup/solve summary
  -h, --help               Print this message
""")
end

function parse_args(args)
    params = Params(nothing, (10, 10, 10), (1.0, 1.0, 1.0), (1, 1, 1), 7, 5, 1, String[])
    i = 1
    while i <= length(args)
        arg = args[i]
        if arg in ("-h", "--help")
            usage()
            exit(0)
        elseif arg in ("-i", "--input")
            i += 1
            i <= length(args) || error("$arg requires a file")
            params.input = args[i]
        elseif arg in ("-a", "--args")
            params.hypredrv_args = args[i:end]
            break
        elseif arg == "-n"
            i + 3 <= length(args) || error("-n requires three values")
            params.n = (parse(Int, args[i + 1]), parse(Int, args[i + 2]), parse(Int, args[i + 3]))
            i += 3
        elseif arg == "-c"
            i + 3 <= length(args) || error("-c requires three values")
            params.c = (parse(Float64, args[i + 1]), parse(Float64, args[i + 2]), parse(Float64, args[i + 3]))
            i += 3
        elseif arg == "-P"
            i + 3 <= length(args) || error("-P requires three values")
            params.p = (parse(Int, args[i + 1]), parse(Int, args[i + 2]), parse(Int, args[i + 3]))
            i += 3
        elseif arg == "-s"
            i += 1
            i <= length(args) || error("-s requires a stencil value")
            params.stencil = parse(Int, args[i])
        elseif arg in ("-ns", "--nsolve")
            i += 1
            i <= length(args) || error("$arg requires a count")
            params.nsolve = parse(Int, args[i])
        elseif arg in ("-v", "--verbose")
            i += 1
            i <= length(args) || error("$arg requires a value")
            params.verbose = parse(Int, args[i])
        else
            error("unknown argument: $arg")
        end
        i += 1
    end
    return params
end

function starts_for(n::Int, p::Int)
    base = div(n, p)
    rest = n - base * p
    return [base * j + min(j, rest) for j in 0:p]
end

function rank_to_coords(rank::Int, p::NTuple{3,Int})
    px, py, pz = p
    x = div(rank, py * pz)
    rem1 = rank - x * py * pz
    y = div(rem1, pz)
    z = rem1 - y * pz
    return (x, y, z)
end

function owner_coord(coord0::Int, starts::Vector{Int})
    return searchsortedlast(starts, coord0) - 1
end

function block_index(g::NTuple{3,Int}, starts)
    return (owner_coord(g[1], starts[1]), owner_coord(g[2], starts[2]), owner_coord(g[3], starts[3]))
end

function block_size(b::NTuple{3,Int}, starts)
    bx, by, bz = b
    sx, sy, sz = starts
    lx = sx[bx + 2] - sx[bx + 1]
    ly = sy[by + 2] - sy[by + 1]
    lz = sz[bz + 2] - sz[bz + 1]
    return lx * ly * lz
end

function block_bases(p::NTuple{3,Int}, starts)
    bases = Array{Int,3}(undef, p...)
    next_base = 0
    for bx in 0:p[1]-1, by in 0:p[2]-1, bz in 0:p[3]-1
        bases[bx + 1, by + 1, bz + 1] = next_base
        next_base += block_size((bx, by, bz), starts)
    end
    return bases
end

function grid_to_index(g::NTuple{3,Int}, b::NTuple{3,Int}, starts, bases)
    bx, by, bz = b
    sx, sy, sz = starts
    lx = sx[bx + 2] - sx[bx + 1]
    ly = sy[by + 2] - sy[by + 1]
    return bases[bx + 1, by + 1, bz + 1] +
           ((g[3] - sz[bz + 1]) * ly + (g[2] - sy[by + 1])) * lx +
           (g[1] - sx[bx + 1])
end

function check_global_indexing(params::Params)
    starts = (starts_for(params.n[1], params.p[1]),
              starts_for(params.n[2], params.p[2]),
              starts_for(params.n[3], params.p[3]))
    bases = block_bases(params.p, starts)
    total = prod(params.n)
    seen = falses(total)
    for bx in 0:params.p[1]-1, by in 0:params.p[2]-1, bz in 0:params.p[3]-1
        block = (bx, by, bz)
        x0, x1 = starts[1][bx + 1], starts[1][bx + 2] - 1
        y0, y1 = starts[2][by + 1], starts[2][by + 2] - 1
        z0, z1 = starts[3][bz + 1], starts[3][bz + 2] - 1
        for z in z0:z1, y in y0:y1, x in x0:x1
            index = grid_to_index((x, y, z), block, starts, bases)
            0 <= index < total || error("global row index $index is out of range")
            !seen[index + 1] || error("duplicate global row index $index")
            seen[index + 1] = true
        end
    end
    all(seen) || error("global row indexing has gaps")
    return nothing
end

struct StencilOffset
    di::Int
    dj::Int
    dk::Int
    weight::Float64
end

function stencil_offsets(stencil::Int)
    offsets = StencilOffset[]
    append!(offsets, [StencilOffset(-1, 0, 0, 1.0), StencilOffset(1, 0, 0, 1.0),
                      StencilOffset(0, -1, 0, 1.0), StencilOffset(0, 1, 0, 1.0),
                      StencilOffset(0, 0, -1, 1.0), StencilOffset(0, 0, 1, 1.0)])
    if stencil >= 19
        for di in -1:1, dj in -1:1, dk in -1:1
            distance = abs(di) + abs(dj) + abs(dk)
            if distance == 2 || (stencil == 27 && distance == 3)
                push!(offsets, StencilOffset(di, dj, dk, 0.25 / distance))
            end
        end
    end
    return offsets
end

function build_local_laplacian(params::Params, rank::Int)
    params.stencil in (7, 19, 27) || error("supported Julia stencils are 7, 19, and 27")
    all(params.n .>= 2) || error("all grid dimensions must be at least 2")
    starts = (starts_for(params.n[1], params.p[1]),
              starts_for(params.n[2], params.p[2]),
              starts_for(params.n[3], params.p[3]))
    bases = block_bases(params.p, starts)
    coords = rank_to_coords(rank, params.p)
    x0, x1 = starts[1][coords[1] + 1], starts[1][coords[1] + 2] - 1
    y0, y1 = starts[2][coords[2] + 1], starts[2][coords[2] + 2] - 1
    z0, z1 = starts[3][coords[3] + 1], starts[3][coords[3] + 2] - 1

    row_start = grid_to_index((x0, y0, z0), coords, starts, bases)
    local_n = (x1 - x0 + 1) * (y1 - y0 + 1) * (z1 - z0 + 1)
    indptr = Int64[0]
    cols = Int64[]
    vals = Float64[]
    rhs = Float64[]
    offsets = stencil_offsets(params.stencil)
    cx, cy, cz = params.c
    coeff_scale = (cx, cy, cz)

    for z in z0:z1, y in y0:y1, x in x0:x1
        boundary = x == 0 || x == params.n[1] - 1 || y == 0 || y == params.n[2] - 1 || z == 0 || z == params.n[3] - 1
        if boundary
            push!(cols, grid_to_index((x, y, z), coords, starts, bases))
            push!(vals, 1.0)
            push!(rhs, y == 0 ? 1.0 : 0.0)
            push!(indptr, length(cols))
            continue
        end

        diag = 0.0
        for offset in offsets
            ng = (x + offset.di, y + offset.dj, z + offset.dk)
            nb = block_index(ng, starts)
            direction_coeff = abs(offset.di) * coeff_scale[1] +
                              abs(offset.dj) * coeff_scale[2] +
                              abs(offset.dk) * coeff_scale[3]
            coeff = offset.weight * direction_coeff
            push!(cols, grid_to_index(ng, nb, starts, bases))
            push!(vals, -coeff)
            diag += coeff
        end
        push!(cols, grid_to_index((x, y, z), coords, starts, bases))
        push!(vals, diag)
        push!(rhs, 0.0)
        push!(indptr, length(cols))
    end

    return row_start, indptr, cols, vals, rhs
end

function options_from(params::Params)
    params.input !== nothing && return read(params.input, String)
    return hypredrive_options(solver=:pcg,
                              preconditioner=:amg,
                              pcg=(max_iter=200, relative_tol=1.0e-8),
                              amg=(print_level=0,))
end

function main()
    params = parse_args(ARGS)
    rank = hypredrive_mpi_world_rank()
    nprocs = hypredrive_mpi_world_size()

    prod(params.p) == nprocs || error("number of MPI ranks ($nprocs) must equal P product $(prod(params.p))")
    all(params.p .<= params.n) || error("processor grid dimensions must not exceed grid dimensions")
    if rank == 0 && get(ENV, "HYPREDRV_JULIA_CHECK_INDEXING", "0") == "1"
        check_global_indexing(params)
    end

    if rank == 0 && (params.verbose & 0x1) != 0
        println()
        println("=====================================================")
        println("              Julia Laplacian Problem Setup")
        println("=====================================================")
        println("Grid dimensions:      $(params.n[1]) x $(params.n[2]) x $(params.n[3])")
        println("Processor topology:   $(params.p[1]) x $(params.p[2]) x $(params.p[3])")
        println("Diffusion coeffs:     $(params.c)")
        println("Discretization:       $(params.stencil)-point stencil")
        println("Number of solves:     $(params.nsolve)")
        println("=====================================================")
        println()
    end

    row_start, indptr, cols, vals, rhs = build_local_laplacian(params, rank)
    if isempty(params.hypredrv_args)
        local_x, info = hypredrive_solve_mpi_csr(indptr, cols, vals, rhs, row_start;
                                                 options=options_from(params),
                                                 nsolve=params.nsolve, comm=:world)
    else
        # input_args[1] carries the configuration; normalize it the same way the
        # options= path does (e.g. inject a quiet general.statistics default).
        input_args = String[hypredrive_options(options_from(params))]
        append!(input_args, params.hypredrv_args)
        local_x, info = hypredrive_solve_mpi_csr(indptr, cols, vals, rhs, row_start;
                                                 options="", input_args=input_args,
                                                 nsolve=params.nsolve, comm=:world)
    end
    local_norm2 = dot(local_x, local_x)
    global_norm = sqrt(hypredrive_mpi_world_sum(local_norm2))

    if rank == 0 && (params.verbose & 0x1) != 0
        println("Iterations:           ", info.iterations)
        println("Setup time:           ", info.setup_time)
        println("Solve time:           ", info.solve_time)
        println("Global solution norm: ", global_norm)
    end
end

main()

File: interfaces/cpp/examples/laplacian.cpp
/******************************************************************************
 * Copyright (c) 2024 Lawrence Livermore National Security, LLC and other
 * HYPRE Project Developers. See the top-level COPYRIGHT file for details.
 *
 * SPDX-License-Identifier: MIT
 ******************************************************************************/

#include <hypredrive.hpp>

#include <algorithm>
#include <array>
#include <cstdlib>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

namespace hd = hypredrive;

namespace
{

struct params
{
   std::array<int, 3> n{16, 16, 16};
   std::array<int, 3> p{1, 1, 1};
   int                stencil = 7;
   int                verbose = 1;
   std::string        input_file{};
   int                hypredrv_argc = 0;       /* override args (incl. -a) */
   char             **hypredrv_argv = nullptr; /* override args, starting at -a */
};

void
usage(const char *argv0)
{
   std::cerr << "Usage: mpiexec -n <np> " << argv0
             << " [-n NX NY NZ] [-P PX PY PZ] [-s 7] [-v LEVEL] [-i YAML]"
                " [-a --path:to:key value ...]\n";
}

params
parse_args(int argc, char **argv)
{
   params out;
   for (int i = 1; i < argc; ++i)
   {
      const std::string arg = argv[i];
      if (arg == "-h" || arg == "--help")
      {
         usage(argv[0]);
         std::exit(EXIT_SUCCESS);
      }
      else if (arg == "-n" && i + 3 < argc)
      {
         out.n = {std::atoi(argv[++i]), std::atoi(argv[++i]), std::atoi(argv[++i])};
      }
      else if (arg == "-P" && i + 3 < argc)
      {
         out.p = {std::atoi(argv[++i]), std::atoi(argv[++i]), std::atoi(argv[++i])};
      }
      else if (arg == "-s" && i + 1 < argc)
      {
         out.stencil = std::atoi(argv[++i]);
      }
      else if (arg == "-v" && i + 1 < argc)
      {
         out.verbose = std::atoi(argv[++i]);
      }
      else if ((arg == "-i" || arg == "--input") && i + 1 < argc)
      {
         out.input_file = argv[++i];
      }
      else if (arg == "-a" || arg == "--args")
      {
         out.hypredrv_argc = argc - i;
         out.hypredrv_argv = argv + i;
         break;
      }
      else
      {
         usage(argv[0]);
         throw std::runtime_error("invalid command-line arguments");
      }
   }
   if (out.stencil != 7)
   {
      throw std::runtime_error("the C++ Laplacian example currently supports only -s 7");
   }
   if (out.n[0] <= 0 || out.n[1] <= 0 || out.n[2] <= 0 || out.p[0] <= 0 ||
       out.p[1] <= 0 || out.p[2] <= 0)
   {
      throw std::runtime_error("grid and processor dimensions must be positive");
   }
   return out;
}

hd::bigint_vector
split_points(int n, int p)
{
   hd::bigint_vector points(static_cast<std::size_t>(p + 1));
   const int         base = n / p;
   const int         rem  = n % p;
   for (int i = 0; i <= p; ++i)
   {
      points[static_cast<std::size_t>(i)] = base * i + std::min(i, rem);
   }
   return points;
}

int
owner_coord(hd::bigint x, const hd::bigint_vector &starts)
{
   return static_cast<int>(std::upper_bound(starts.begin(), starts.end(), x) -
                           starts.begin()) -
          1;
}

hd::bigint
block_size(const std::array<hd::bigint_vector, 3> &starts,
           const std::array<int, 3>               &coords)
{
   hd::bigint total = 1;
   for (int d = 0; d < 3; ++d)
   {
      total *= starts[d][static_cast<std::size_t>(coords[d] + 1)] -
               starts[d][static_cast<std::size_t>(coords[d])];
   }
   return total;
}

hd::bigint
local_id(const std::array<hd::bigint_vector, 3> &starts, const std::array<int, 3> &block,
         hd::bigint x, hd::bigint y, hd::bigint z)
{
   const auto x0 = starts[0][static_cast<std::size_t>(block[0])];
   const auto y0 = starts[1][static_cast<std::size_t>(block[1])];
   const auto z0 = starts[2][static_cast<std::size_t>(block[2])];
   const auto nx = starts[0][static_cast<std::size_t>(block[0] + 1)] - x0;
   const auto ny = starts[1][static_cast<std::size_t>(block[1] + 1)] - y0;
   return ((z - z0) * ny + (y - y0)) * nx + (x - x0);
}

struct grid
{
   MPI_Comm                         cart = MPI_COMM_NULL;
   std::array<int, 3>               n{};
   std::array<int, 3>               p{};
   std::array<int, 3>               coords{};
   std::array<hd::bigint_vector, 3> starts{};
   hd::bigint_vector                rank_offsets{};

   ~grid()
   {
      if (cart != MPI_COMM_NULL) MPI_Comm_free(&cart);
   }

   hd::bigint
   block_offset(const std::array<int, 3> &block) const
   {
      int rank = 0;
      int c[3] = {block[0], block[1], block[2]};
      MPI_Cart_rank(cart, c, &rank);
      return rank_offsets[static_cast<std::size_t>(rank)];
   }

   hd::bigint
   global_id(hd::bigint x, hd::bigint y, hd::bigint z) const
   {
      const std::array<int, 3> block{owner_coord(x, starts[0]), owner_coord(y, starts[1]),
                                     owner_coord(z, starts[2])};
      return block_offset(block) + local_id(starts, block, x, y, z);
   }

   hd::bigint
   row_start() const
   {
      return block_offset(coords);
   }
   hd::bigint
   local_size() const
   {
      return block_size(starts, coords);
   }
   hd::bigint
   row_end() const
   {
      return row_start() + local_size() - 1;
   }
};

grid
make_grid(MPI_Comm comm, const params &par)
{
   int size = 1;
   MPI_Comm_size(comm, &size);
   if (par.p[0] * par.p[1] * par.p[2] != size)
   {
      throw std::runtime_error("-P product must match MPI size");
   }
   grid g;
   g.n            = par.n;
   g.p            = par.p;
   int dims[3]    = {par.p[0], par.p[1], par.p[2]};
   int periods[3] = {0, 0, 0};
   MPI_Cart_create(comm, 3, dims, periods, 0, &g.cart);
   int rank = 0;
   MPI_Comm_rank(g.cart, &rank);
   int coords[3] = {0, 0, 0};
   MPI_Cart_coords(g.cart, rank, 3, coords);
   g.coords = {coords[0], coords[1], coords[2]};
   for (int d = 0; d < 3; ++d) g.starts[d] = split_points(g.n[d], g.p[d]);
   g.rank_offsets.assign(static_cast<std::size_t>(size), 0);
   hd::bigint_vector rank_sizes(static_cast<std::size_t>(size), 0);
   for (int r = 0; r < size; ++r)
   {
      int rc[3] = {0, 0, 0};
      MPI_Cart_coords(g.cart, r, 3, rc);
      rank_sizes[static_cast<std::size_t>(r)] =
         block_size(g.starts, {rc[0], rc[1], rc[2]});
   }
   for (int r = 1; r < size; ++r)
   {
      g.rank_offsets[static_cast<std::size_t>(r)] =
         g.rank_offsets[static_cast<std::size_t>(r - 1)] +
         rank_sizes[static_cast<std::size_t>(r - 1)];
   }
   return g;
}

constexpr const char *default_options_yaml = R"yaml(
general:
  name: cpp-laplacian
  statistics: 1
solver:
  pcg:
    max_iter: 100
    relative_tol: 1.0e-8
    print_level: 0
preconditioner:
  amg:
    print_level: 0
)yaml";

void
build_local_csr(const grid &g, hd::bigint_vector &indptr, hd::bigint_vector &cols,
                hd::real_vector &vals, hd::real_vector &rhs)
{
   const auto xs = g.starts[0][static_cast<std::size_t>(g.coords[0])];
   const auto xe = g.starts[0][static_cast<std::size_t>(g.coords[0] + 1)];
   const auto ys = g.starts[1][static_cast<std::size_t>(g.coords[1])];
   const auto ye = g.starts[1][static_cast<std::size_t>(g.coords[1] + 1)];
   const auto zs = g.starts[2][static_cast<std::size_t>(g.coords[2])];
   const auto ze = g.starts[2][static_cast<std::size_t>(g.coords[2] + 1)];
   indptr.assign(static_cast<std::size_t>(g.local_size() + 1), 0);
   rhs.assign(static_cast<std::size_t>(g.local_size()), 1.0);
   cols.clear();
   vals.clear();
   hd::bigint row   = 0;
   const int  dx[6] = {-1, 1, 0, 0, 0, 0};
   const int  dy[6] = {0, 0, -1, 1, 0, 0};
   const int  dz[6] = {0, 0, 0, 0, -1, 1};
   for (auto z = zs; z < ze; ++z)
   {
      for (auto y = ys; y < ye; ++y)
      {
         for (auto x = xs; x < xe; ++x)
         {
            hd::real diag = 0.0;
            for (int q = 0; q < 6; ++q)
            {
               const auto xn = x + dx[q];
               const auto yn = y + dy[q];
               const auto zn = z + dz[q];
               if (xn >= 0 && xn < g.n[0] && yn >= 0 && yn < g.n[1] && zn >= 0 &&
                   zn < g.n[2])
               {
                  cols.push_back(g.global_id(xn, yn, zn));
                  vals.push_back(-1.0);
                  diag += 1.0;
               }
            }
            cols.push_back(g.global_id(x, y, z));
            vals.push_back(diag);
            indptr[static_cast<std::size_t>(++row)] =
               static_cast<hd::bigint>(cols.size());
         }
      }
   }
}

} // namespace

int
main(int argc, char **argv)
{
   int ierr = EXIT_SUCCESS;
   int rank = 0;
   MPI_Init(&argc, &argv);
   MPI_Comm_rank(MPI_COMM_WORLD, &rank);

   try
   {
      const params      par = parse_args(argc, argv);
      grid              g   = make_grid(MPI_COMM_WORLD, par);
      hd::bigint_vector indptr, cols;
      hd::real_vector   vals, rhs;
      build_local_csr(g, indptr, cols, vals, rhs);

      hd::initialize();
      hd::driver drv(g.cart);
      drv.set_library_mode();
      const auto solver_options =
         par.input_file.empty() ? std::string(default_options_yaml) : par.input_file;
      const std::vector<std::string> overrides(par.hypredrv_argv,
                                               par.hypredrv_argv + par.hypredrv_argc);
      drv.parse_yaml(solver_options, overrides);
      drv.set_matrix_from_csr(g.row_start(), g.row_end(), indptr.data(), cols.data(),
                              vals.data());
      drv.set_rhs_from_array(g.row_start(), g.row_end(), rhs.data());
      drv.set_zero_initial_guess();
      drv.solve();
      const auto global_unknowns =
         static_cast<hd::bigint>(par.n[0]) * par.n[1] * par.n[2];
      if (rank == 0 && par.verbose > 0)
      {
         std::cout << "ranks          : " << par.p[0] * par.p[1] * par.p[2] << "\n"
                   << "unknowns       : " << global_unknowns << "\n"
                   << "solution norm  : " << drv.get_solution_norm("l2") << "\n"
                   << "iterations     : " << drv.get_linear_solver_num_iter() << "\n";
      }
      if (rank == 0) drv.print_stats();
      hd::finalize();
   }
   catch (const std::exception &e)
   {
      int rank = 0;
      MPI_Comm_rank(MPI_COMM_WORLD, &rank);
      if (rank == 0) std::cerr << e.what() << "\n";
      // Best-effort cleanup; ignore finalization errors while reporting the original
      // failure.
      try
      {
         hd::finalize();
      }
      catch (const std::exception &)
      {
      }
      ierr = EXIT_FAILURE;
   }
   MPI_Finalize();
   return ierr;
}
